helios-services/src/main/java/com/spotify/helios/agent/QueueingHistoryWriter.java

```diff
@@ -130,6 +158,11 @@ protected void startUp() throws Exception {
   protected void shutDown() throws Exception {
     zkWriterExecutor.shutdownNow();
     zkWriterExecutor.awaitTermination(1, TimeUnit.MINUTES);
+
+    if (kafkaProducer.isPresent()) {
+      // Otherwise it enters an infinite loop for some reason.
```

Comment:
Docs say that:
    The producer manages a single background thread that does I/O as well as a TCP connection to each of the brokers it needs to communicate with. Failure to close the producer after use will leak these resources.
So I guess maybe it's because the background thread is never stopped or something like that.